### PR TITLE
CRYPTO-60: opensslCipher support GCM mode

### PR DESCRIPTION
**Pull request by kexianda**
[CRYPTO-60](https://issues.apache.org/jira/browse/CRYPTO-60)
support GCM mode

### CODE REVIEW

File: src/main/java/org/apache/commons/crypto/cipher/OpenSslCommonMode.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    @Override
+    public int doFinal(byte[] input, int inputOffset, int inputLen, byte[] output, int outputOffset)
+            throws ShortBufferException, IllegalBlockSizeException, BadPaddingException {
+        checkState();
+
+        int len = OpenSslNative.updateByteArray(context, input, inputOffset,
+                inputLen, output, outputOffset, output.length - outputOffset);
+
+        len += OpenSslNative.doFinalByteArray(context, output, outputOffset + len,
+                output.length - outputOffset);
```
Comment (sundapeng) [inline src/main/java/org/apache/commons/crypto/cipher/OpenSslCommonMode.java]: Should `output.length - outputOffset` be `output.length - outputOffset - len` ?

File: src/main/native/org/apache/commons/crypto/org_apache_commons_crypto.h [hunk 2]
```diff
@@ -216,12 +216,15 @@ static FARPROC WINAPI do_dlsym(JNIEnv *env, HMODULE handle, LPCSTR symbol) {
 #define ENCRYPT_MODE 1
 #define DECRYPT_MODE 0
 
-/** Currently only support AES/CTR/NoPadding. */
+/** Currently only support AES/CTR/NoPadding, AES/CBC/NoPadding. AES/GCM/NoPadding */
```
Comment (sundapeng) [inline src/main/native/org/apache/commons/crypto/org_apache_commons_crypto.h]: AES/CTR/NoPadding, AES/CBC/NoPadding, AES/CBC/PKCS5Padding,  AES/GCM/NoPadding

File: src/main/java/org/apache/commons/crypto/cipher/OpenSslGaloisCounterMode.java [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+            tag.put(input, input.length - getTagLen(), getTagLen());
+            tag.flip();
+            evpCipherCtxCtrl(context, OpenSslEvpCtrlValues.AEAD_SET_TAG.getValue(), getTagLen(), tag);
+        } else {
+            len = OpenSslNative.updateByteArray(context, input, inputOffset,
+                    inputLen, output, outputOffset, output.length - outputOffset);
+        }
+
+        len += OpenSslNative.doFinalByteArray(context, output, outputOffset + len,
+                output.length - outputOffset);
```
Comment (sundapeng) [inline src/main/java/org/apache/commons/crypto/cipher/OpenSslGaloisCounterMode.java]: should `output.length - outputOffset ` be `output.length - outputOffset - len`

**Comment by yaronlev171** [comment]
Hi, I am very interested in this PR - any reason why it was not merged? 
Thanks,
Yaron.

**Comment by kexianda** [comment]
@yaronlev171,  This PR implements GCM  only in Cipher level.  it needs to be tested more, but I was busy on other stuffs. I will be back on this PR soon, i will refine the code and do more testing.

now,  Stream level api does not support GCM. We need to do more work on stream API.

**Comment by yaronlev171** [comment]
@kexianda thanks for the reply.
Do you know of any standard/mainstream alternative for getting gcm to work fully with aes-ni?

File: src/main/java/org/apache/commons/crypto/cipher/OpenSslGaloisCounterMode.java [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+ * This class do the real work(Encryption/Decryption/Authentication) for the authenticated mode: GCM.
+ *
+ * It calls the OpenSSL API to implement the JCE-like behavior
+ *
+ * @since 1.1
+ */
+class OpenSslGaloisCounterMode extends OpenSslFeedbackCipher {
+
+    // buffer for AAD data; if consumed, set as null
+    private ByteArrayOutputStream aadBuffer = new ByteArrayOutputStream();
```
Comment (yaronlev171) [inline src/main/java/org/apache/commons/crypto/cipher/OpenSslGaloisCounterMode.java]: this should be set inside init() so the cipher with aad is reusable after clean(). [tested]

Comment (kexianda) [inline src/main/java/org/apache/commons/crypto/cipher/OpenSslGaloisCounterMode.java]: thank you for this comment.   I'll update the patch.

File: src/main/java/org/apache/commons/crypto/cipher/OpenSsl.java [hunk 5]
```diff
@@ -317,53 +350,43 @@ public int update(byte[] input, int inputOffset, int inputLen,
...
-            OpenSslNative.clean(context);
-            context = 0;
+        if (opensslBlockCipher != null) {
+            opensslBlockCipher.clean();
         }
     }
 
     /** Checks whether context is initialized. */
     private void checkState() {
-        Utils.checkState(context != 0);
```
Comment (kexianda) [inline src/main/java/org/apache/commons/crypto/cipher/OpenSsl.java]: 1. behavior is not changed.  The context was moved to to the Cipher(OpenSslFeedbackCipher).
The cipher will check its own state.

2. Well, I am considering to remove OpenSsl.java.  From my point of view, now this layer(OpenSsl.java) seems redundant , we can move the logic(paramenters check) in OpenSsl.java to OpenSslCipher.  
Then,  the codebase will be clear:
* OpenSslCipher which implements the CryptoCipher interface which is exposed to the users.
* OpenSslFeedbackCipher and its sub classes (OpenSslGaloisCounterMode,OpenSslCommonMode) are private classes. they do the real work(encription/decription) for different modes.
* OpenSslNative: JNI class
Maybe we can do this later.

**Comment by sundapeng** [comment]
+1, thank @kexianda for the contribution!

**Comment by kexianda** [comment]
Thank @sundapeng.  I will update the testcases  with samples for GCM/GMAC soon.

**Comment by kexianda** [comment]
BUILD SUCCESS on my local dev enviroment. but failed on travis-ci.   
weird...

**Comment by sundapeng** [comment]
Thank @kexianda, I will fix it.

**Comment by laxmanprabhu** [comment]
Is there a timeline for the release of this feature. The version is 3 years old.

**Pull request closed without merging**